solr/core/src/java/org/apache/solr/cloud/api/collections/CreateCollectionCmd.java

```diff
@@ -407,14 +466,81 @@ public void call(ClusterState clusterState, ZkNodeProps message, NamedList resul
           .assignPullReplicas(numPullReplicas)
           .onNodes(nodeList)
           .build();
-      Assign.AssignStrategyFactory assignStrategyFactory = new Assign.AssignStrategyFactory(cloudManager);
+      // if Strategy.POLICY is in use AND maxShardsPerNode was specified then for back-compat it needs to
+      // be translated into a rule (see SOLR-12847).
+      // Modify the autoscaling rules as needed BEFORE actually calling assign, setting configToRestore as
+      // a side-effect to restore the original config if the creation fails
+      if ((strategy == Assign.AssignStrategyFactory.Strategy.POLICY) &&
+          (maxReplicasPerNode != null && maxReplicasPerNode != Integer.MAX_VALUE)) {
+        maybeAddMaxReplicasRule(cloudManager, maxReplicasPerNode, docCollection, configToRestore);
```

Comment:
I would have preferred to do this outside of this method instead of adding a side effect here. That way the addition of the clause as well as its cleanup can be in one place.